pkg/services/accesscontrol/accesscontrolmock/mock.go

```diff
@@ -0,0 +1,153 @@
+package accesscontrolmock
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+)
+
+type fullAccessControl interface {
+	accesscontrol.AccessControl
+	GetUserBuiltInRoles(user *models.SignedInUser) []string
+	RegisterFixedRoles() error
+}
+
+type Calls struct {
+	Evaluate            []interface{}
+	GetUserPermissions  []interface{}
+	IsDisabled          []interface{}
+	DeclareFixedRoles   []interface{}
+	GetUserBuiltInRoles []interface{}
+	RegisterFixedRoles  []interface{}
+}
+
+type Mock struct {
+	// Unless an override is provided, permissions will be returned by GetUserPermissions
+	permissions []*accesscontrol.Permission
+	// Unless an override is provided, disabled will be returned by IsDisabled
+	disabled bool
+	// Unless an override is provided, builtInRoles will be returned by GetUserBuiltInRoles
+	builtInRoles []string
+
+	// Track the list of calls
+	Calls *Calls
+
+	// Override functions
+	EvaluateFunc            func(context.Context, *models.SignedInUser, accesscontrol.Evaluator) (bool, error)
+	GetUserPermissionsFunc  func(context.Context, *models.SignedInUser) ([]*accesscontrol.Permission, error)
+	IsDisabledFunc          func() bool
+	DeclareFixedRolesFunc   func(...accesscontrol.RoleRegistration) error
+	GetUserBuiltInRolesFunc func(user *models.SignedInUser) []string
+	RegisterFixedRolesFunc  func() error
+}
+
+type MockOptions func(*Mock)
+
+// Ensure the mock stays in line with the interface
+var _ fullAccessControl = New()
+
+func New() *Mock {
+	calls := Calls{}
+	mock := &Mock{
+		Calls:        &calls,
+		disabled:     false,
+		permissions:  []*accesscontrol.Permission{},
+		builtInRoles: []string{},
+	}
+
+	return mock
+}
+
+func (m *Mock) WithPermissions(permissions []*accesscontrol.Permission) *Mock {
```

Comment:
At the beginning, I did this with the functional options pattern but it gave really long calls outside of the package.

Reply:
Why'd that make longer calls? 🤔

Reply:
When I used the functional options pattern I had:
```go
type MockOption func(*Mock)

func WithPermissions(permissions []*accesscontrol.Permission) MockOption {
  return func(m *Mock) {
    m.permissions = permissions
  }
}
```
Outside of this package, I'd have for instance:
```go
mock := accesscontrolmock.New(accesscontrolmock.WithPermissions(permissions), accesscontrolmock.WithDisabled())
```
So it was long due to the repetition of the package name mostly. And for table tests where I'd typically inline the permissions list definition, it was even bigger.

Reply:
Ah, I see.

Reply:
Would it be possible to not have these as pointer receivers so that say

```
a := New().WithPermissions(permissions)
b := a.WithDisabled()
```

doesn't disable a?

Reply:
Ok sure, I just need to make `Calls` not a pointer anymore :+1: